{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# Sequence Extraction Using Pyfaidx\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_count_threshold = 50;\n",
    "NSUN6_NSUN2_cutoff_factor = 0.6; #arbitrary at the moment, between 0 and 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Imports & Path Set-up\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "import math\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pyfaidx import Fasta\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import colors\n",
    "import logomaker as lm\n",
    "\n",
    "PATH = '../GenomeData/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Functions\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_seq_genome(genome, chrom, start, end, strand = '+'):\n",
    "    '''\n",
    "    ath: tair10\n",
    "    cel: cel235\n",
    "    hsa: hg38\n",
    "    mmu: mm10\n",
    "\n",
    "    Usage:\n",
    "        - If strand == '+', return sequence\n",
    "        - If strand == '-', return sequence revese complementary\n",
    "        - With hg38.get_seq(chrom, start, end, rc= strand == '-').seq get correct position of miRNA,\n",
    "        for mapping result, you should modify by yourself.\n",
    "    '''\n",
    "    \n",
    "    if genome == 'cel':\n",
    "        genome = 'cel235'\n",
    "    elif genome == 'hsa':\n",
    "        genome = 'hg38'\n",
    "    elif genome == 'mmu':\n",
    "        genome = 'mm10'\n",
    "    elif genome == 'ath':\n",
    "        genome = 'tair10'\n",
    "    elif genome == 'mml':\n",
    "        genome = 'Macaca_mulatta.Mmul_10.dna.toplevel'\n",
    "    elif genome == 'dre':\n",
    "        genome = 'Danio_rerio.GRCz11.dna.primary_assembly'\n",
    "    elif genome == 'dme':\n",
    "        genome = 'Drosophila_melanogaster.BDGP6.32.dna.toplevel'\n",
    "\n",
    "\n",
    "    gene = Fasta(f\"{PATH}/{genome}.fa\") \n",
    "    \n",
    "    #hg38.get_seq('chr1', 100009588 + 1, 100009608, rc=False).seq\n",
    "    return(gene.get_seq(chrom, start+1, end, rc= strand == '-').seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Preprocessed Data Handling\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The provided data needs to be read into a pandas data frame for further processing..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ubsseq_data_raw = pd.read_excel(\"ubsseq_preprocessed_data.xlsx\")\n",
    "ubsseq_data_raw_shape = ubsseq_data_raw.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The data needs to be filtered to remove genes that were not sampled enough to be statistically significant, this is determined by the unconverted/converted attribute. If the combined sum of these attributes is less than ??? then the gene has not shown enough influence from methylation to be considered."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ubsseq_data_sample_count = ubsseq_data_raw[['unconverted', 'converted']].sum(axis=1)\n",
    "ubsseq_data_significant = ubsseq_data_raw[ubsseq_data_sample_count > sample_count_threshold]\n",
    "ubsseq_data_significant = ubsseq_data_significant[ubsseq_data_significant['chrom'] != 'MT']\n",
    "ubsseq_data_significant_shape = ubsseq_data_significant.shape\n",
    "print(f\"Filtered data contains {ubsseq_data_significant_shape[0]} genes out of an initial {ubsseq_data_raw_shape[0]} genes, a {ubsseq_data_significant_shape[0]/ubsseq_data_raw_shape[0]*100:.2f}% retention\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Ratio Control Selection\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A scatter graph will show the correlation between the control ratio and NSUN6 ratio to help determine which genes should be selected for the positive data set. \n",
    "- Can do both a strict and relax cut-off."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(ubsseq_data_significant['ratio_siControl'], ubsseq_data_significant['ratio_siNSUN6'], label='Data Points')\n",
    "plt.xlabel('siControl Ratio')\n",
    "plt.ylabel('siNSUN6 ratio')\n",
    "\n",
    "a, b = np.polyfit(ubsseq_data_significant['ratio_siControl'], ubsseq_data_significant['ratio_siNSUN6'], 1)\n",
    "\n",
    "# Not sure how to adjust, come back to this later\n",
    "a = a - NSUN6_NSUN2_cutoff_factor\n",
    "\n",
    "plt.plot(ubsseq_data_significant['ratio_siControl'], a * np.array(ubsseq_data_significant['ratio_siControl']) + b, color='red', label='Line of Best Fit')\n",
    "\n",
    "# Joel's Test for Understanding\n",
    "#gene_search = ubsseq_data_significant[ubsseq_data_significant['gene_name'] == \"ENSG00000083444(PLOD1)\"]\n",
    "gene_search = ubsseq_data_significant[ubsseq_data_significant['delta_ratio'] > (1-a)]\n",
    "plt.scatter(gene_search['ratio_siControl'], gene_search['ratio_siNSUN6'], label='Gene Search')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "delta ratio will be (control - NSUN6) / control\n",
    "\n",
    "!!! Need to adjust threshold to ensure more accurate reading of CTCCA motif comes through, otherwise NSUN2 modification is too present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"Best fit gradient is {a:.2f} and y-intercept is {b:.2f}\")\n",
    "ideal_delta_threshold = 1 - a + b\n",
    "print(f\"Ideal data threshold should be above delta ratio {(ideal_delta_threshold) * 100:.2f}%\")\n",
    "ubsseq_data_NSUN6 = ubsseq_data_significant[ubsseq_data_significant['delta_ratio'] >= ideal_delta_threshold]\n",
    "ubsseq_data_NSUN6_shape = ubsseq_data_NSUN6.shape\n",
    "print(f\"This threshold includes {ubsseq_data_NSUN6_shape[0]} samples out of the {ubsseq_data_significant_shape[0]} valid samples, a {ubsseq_data_NSUN6_shape[0]/ubsseq_data_significant_shape[0]*100:.2f}% retention\")\n",
    "\n",
    "#We might want to also explore data we assume not be modified by NSUN6, likely NSUN2\n",
    "ubsseq_data_NSUN2 = ubsseq_data_significant[ubsseq_data_significant['delta_ratio'] < ideal_delta_threshold]\n",
    "\n",
    "### Data clean\n",
    "ubsseq_data_NSUN6=ubsseq_data_NSUN6.drop([\n",
    "    'gene_type','gene_pos','unconverted','converted','ratio_siControl','ratio_siNSUN6','delta_ratio'\n",
    "    ], axis=1)\n",
    "ubsseq_data_NSUN2=ubsseq_data_NSUN2.drop([\n",
    "    'gene_type','gene_pos','unconverted','converted','ratio_siControl','ratio_siNSUN6','delta_ratio'\n",
    "    ], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### distribution of data by delta ratio\n",
    "distribution_data = ubsseq_data_significant['delta_ratio']\n",
    "\n",
    "# N is the count in each bin, bins is the lower-limit of the bin\n",
    "N, bins, patches = plt.hist(distribution_data.to_list(), bins=100, linewidth=0.5, edgecolor=\"white\")\n",
    "\n",
    "# Now, we'll loop through our objects and set the color of each accordingly\n",
    "for thisbin, thispatch in zip(bins, patches):\n",
    "    if(thisbin > ideal_delta_threshold):\n",
    "        color = plt.cm.viridis(0.1)\n",
    "    else:\n",
    "        color = plt.cm.viridis(0.5)\n",
    "    thispatch.set_facecolor(color)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Sequence Extraction\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extend by 50:50, 30:70, 70:30 for each sequence - always length of 101"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Use the samples in df 'ubsseq_data_likely_NSUN6_modified'\n",
    "\n",
    "ubsseq_data_NSUN6['chrom'] = ubsseq_data_NSUN6['chrom'].replace(\n",
    "    [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22, 'X'], \n",
    "    ['chr1','chr2','chr3','chr4','chr5','chr6','chr7','chr8','chr9','chr10','chr11','chr12','chr13','chr14','chr15','chr16','chr17','chr18','chr19','chr20','chr21','chr22','chrX'])\n",
    "\n",
    "#likely_NSUN6_modified_seq_extracted = ubsseq_data_likely_NSUN6_modified;\n",
    "seq_extract = [[],[],[]];\n",
    "\n",
    "for ind in ubsseq_data_NSUN6.index:\n",
    "    gene = ubsseq_data_NSUN6['gene_name'][ind]\n",
    "    chrome = ubsseq_data_NSUN6['chrom'][ind]\n",
    "    strand = ubsseq_data_NSUN6['strand'][ind]\n",
    "    pos = ubsseq_data_NSUN6['pos'][ind]\n",
    "    seq_extract[0].append(extract_seq_genome(\"hg38\", chrome, pos-1-50, pos+50, strand)) #50:50\n",
    "    seq_extract[1].append(extract_seq_genome(\"hg38\", chrome, pos-1-30, pos+70, strand)) #30:70\n",
    "    seq_extract[2].append(extract_seq_genome(\"hg38\", chrome, pos-1-70, pos+30, strand)) #70:30\n",
    "\n",
    "ubsseq_NSUN6_extracted = ubsseq_data_NSUN6.assign(\n",
    "    seq5050=seq_extract[0], seq3070=seq_extract[1], seq7030=seq_extract[2])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The same can be done for the samples likely not modified by NSUN6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ubsseq_data_NSUN2['chrom'] = ubsseq_data_NSUN2['chrom'].replace(\n",
    "    [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22, 'X'], \n",
    "    ['chr1','chr2','chr3','chr4','chr5','chr6','chr7','chr8','chr9','chr10','chr11','chr12','chr13','chr14','chr15','chr16','chr17','chr18','chr19','chr20','chr21','chr22','chrX'])\n",
    "\n",
    "#likely_NSUN6_modified_seq_extracted = ubsseq_data_likely_NSUN6_modified;\n",
    "seq_extract = [[],[],[]];\n",
    "\n",
    "for ind in ubsseq_data_NSUN2.index:\n",
    "    gene = ubsseq_data_NSUN2['gene_name'][ind]\n",
    "    chrome = ubsseq_data_NSUN2['chrom'][ind]\n",
    "    strand = ubsseq_data_NSUN2['strand'][ind]\n",
    "    pos = ubsseq_data_NSUN2['pos'][ind]\n",
    "    seq_extract[0].append(extract_seq_genome(\"hg38\", chrome, pos-1-50, pos+50, strand)) #50:50\n",
    "    seq_extract[1].append(extract_seq_genome(\"hg38\", chrome, pos-1-30, pos+70, strand)) #30:70\n",
    "    seq_extract[2].append(extract_seq_genome(\"hg38\", chrome, pos-1-70, pos+30, strand)) #70:30\n",
    "\n",
    "ubsseq_NSUN2_extracted = ubsseq_data_NSUN2.assign(\n",
    "    seq5050=seq_extract[0], seq3070=seq_extract[1], seq7030=seq_extract[2])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Motif Sequence Logo\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Generate a sequence logo. To do so we focus on the nucleotides immediately adjacent to the methylated 'C' theoretically in index 50. This value should be in the centre of each extended sequence so we will just take the values from 45 to 56. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ubsseq_NSUN6_extracted_5050 = ubsseq_NSUN6_extracted['seq5050'].to_list()\n",
    "for i in range (0, len(ubsseq_NSUN6_extracted_5050)):\n",
    "    ubsseq_NSUN6_extracted_5050[i] = ubsseq_NSUN6_extracted_5050[i][45:56]\n",
    "counts_mat = lm.alignment_to_matrix(ubsseq_NSUN6_extracted_5050)\n",
    "\n",
    "# Counts matrix -> Information matrix\n",
    "info_mat = lm.transform_matrix(counts_mat, from_type='counts', to_type='information')\n",
    "\n",
    "logo_count = lm.Logo(df=counts_mat,\n",
    "               font_name='Arial Rounded MT Bold',\n",
    "               figsize=(10,3)) # Can change to plot entropy or percentage instead of count\n",
    "logo_count.ax.set_xlabel('Position',fontsize=14)\n",
    "logo_count.ax.set_ylabel('Count', labelpad=-1,fontsize=14)\n",
    "\n",
    "logo_info = lm.Logo(df=info_mat,\n",
    "               font_name='Arial Rounded MT Bold',\n",
    "               figsize=(10,3)) # Can change to plot entropy or percentage instead of count\n",
    "logo_info.ax.set_xlabel('Position',fontsize=14)\n",
    "logo_info.ax.set_ylabel('Information (bits)', labelpad=-1,fontsize=14)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The above sequence logo shows that all of the data is correctly centred around the sampled cytosine and also shows the known NSUN6 motif 'CTCCA' (or 'CUCCA' but T<->U).\n",
    "\n",
    "Now we can explore the data that was cut off in a second sequence logo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ubsseq_NSUN2_extracted_5050 = ubsseq_NSUN2_extracted['seq5050'].to_list()\n",
    "for i in range (0, len(ubsseq_NSUN2_extracted_5050)):\n",
    "    ubsseq_NSUN2_extracted_5050[i] = ubsseq_NSUN2_extracted_5050[i][45:56]\n",
    "counts_mat = lm.alignment_to_matrix(ubsseq_NSUN2_extracted_5050)\n",
    "\n",
    "# Counts matrix -> Information matrix\n",
    "info_mat = lm.transform_matrix(counts_mat, from_type='counts', to_type='information')\n",
    "\n",
    "logo_count = lm.Logo(df=counts_mat,\n",
    "               font_name='Arial Rounded MT Bold',\n",
    "               figsize=(10,3)) # Can change to plot entropy or percentage instead of count\n",
    "logo_count.ax.set_xlabel('Position',fontsize=14)\n",
    "logo_count.ax.set_ylabel('Count', labelpad=-1,fontsize=14)\n",
    "\n",
    "logo_info = lm.Logo(df=info_mat,\n",
    "               font_name='Arial Rounded MT Bold',\n",
    "               figsize=(10,3)) # Can change to plot entropy or percentage instead of count\n",
    "logo_info.ax.set_xlabel('Position',fontsize=14)\n",
    "logo_info.ax.set_ylabel('Information (bits)', labelpad=-1,fontsize=14)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Saving Back To Fasta\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add them to a fasta file. Label fasta entries as...\n",
    "- posx-50-50\n",
    "- posx-30-70\n",
    "- posx-70-30"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "NSUN6_pos_fasta = []\n",
    "NSUN6_pos_csv = []\n",
    "for ind in ubsseq_NSUN6_extracted.index:\n",
    "    gene_name = ubsseq_NSUN6_extracted['gene_name'][ind]\n",
    "    NSUN6_pos_fasta.append('>%s_5050\\n%s' % (gene_name, ubsseq_NSUN6_extracted['seq5050'][ind]))\n",
    "    NSUN6_pos_fasta.append('>%s_3070\\n%s' % (gene_name, ubsseq_NSUN6_extracted['seq3070'][ind]))\n",
    "    NSUN6_pos_fasta.append('>%s_7030\\n%s' % (gene_name, ubsseq_NSUN6_extracted['seq7030'][ind]))\n",
    "    NSUN6_pos_csv.append('%s,5050,%s' % (gene_name, ubsseq_NSUN6_extracted['seq5050'][ind]))\n",
    "    NSUN6_pos_csv.append('%s,3070,%s' % (gene_name, ubsseq_NSUN6_extracted['seq3070'][ind]))\n",
    "    NSUN6_pos_csv.append('%s,7030,%s' % (gene_name, ubsseq_NSUN6_extracted['seq7030'][ind]))\n",
    "\n",
    "with open('out/NSUN6_pos_extend.fasta', 'w') as file:\n",
    "     file.write('\\n'.join(NSUN6_pos_fasta))\n",
    "\n",
    "with open('out/NSUN6_pos_extend.csv', 'w') as file:\n",
    "     file.write('\\n'.join(NSUN6_pos_csv))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And then create a Fasta file for our negative dataset..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "NSUN6_neg_fasta = []\n",
    "NSUN6_neg_csv = []\n",
    "for ind in ubsseq_NSUN2_extracted.index:\n",
    "    gene_name = ubsseq_NSUN2_extracted['gene_name'][ind]\n",
    "    NSUN6_neg_fasta.append('>%s_5050\\n%s' % (gene_name, ubsseq_NSUN2_extracted['seq5050'][ind]))\n",
    "    NSUN6_neg_fasta.append('>%s_3070\\n%s' % (gene_name, ubsseq_NSUN2_extracted['seq3070'][ind]))\n",
    "    NSUN6_neg_fasta.append('>%s_7030\\n%s' % (gene_name, ubsseq_NSUN2_extracted['seq7030'][ind]))\n",
    "    NSUN6_neg_csv.append('%s,5050,%s' % (gene_name, ubsseq_NSUN2_extracted['seq5050'][ind]))\n",
    "    NSUN6_neg_csv.append('%s,3070,%s' % (gene_name, ubsseq_NSUN2_extracted['seq3070'][ind]))\n",
    "    NSUN6_neg_csv.append('%s,7030,%s' % (gene_name, ubsseq_NSUN2_extracted['seq7030'][ind]))\n",
    "\n",
    "with open('out/NSUN6_neg_extend.fasta', 'w') as file:\n",
    "     file.write('\\n'.join(NSUN6_neg_fasta))\n",
    "\n",
    "with open('out/NSUN6_neg_extend.csv', 'w') as file:\n",
    "     file.write('\\n'.join(NSUN6_neg_csv))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Calculating Structure\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To copy the fasta file into HPC3 run the following from this pc:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [],
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "scp -r C:\\Users\\joele\\source\\repos\\MLRNAProject\\UBS-SeqDemoTask\\out\\NSUN6_pos_extend.csv jdevans@hpc3.ust.hk:/scratch/PI/tgx/jdevans/RNAFold"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "On the linux system, RNAFold can be used to calculate the structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "python fold.py NSUN6_extend.csv Data_Out/NSUN6_extend_structure.csv\n",
    "#RNAfold <NSUN6_extended.fasta > NSUN6_structure.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And then to download the output run the following from this pc:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "scp -r jdevans@hpc3.ust.hk:/scratch/PI/tgx/jdevans/RNAFold/Data_Out/NSUN6_extend_structure.csv C:\\Users\\joele\\source\\repos\\MLRNAProject\\UBS-SeqDemoTask\\out\\"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now split the returned file into three files for each positional condition (50:50, 30:70, 70:30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_input = pd.read_csv('out/NSUN6_extend_structure.csv', sep = '\\t', engine='python')\n",
    "df_input[df_input['positional'] == 5050].drop('positional', axis=1).to_csv('out/NSUN6_extend_structure_5050.csv', sep = '\\t', header = True, index = False)\n",
    "df_input[df_input['positional'] == 3070].drop('positional', axis=1).to_csv('out/NSUN6_extend_structure_3070.csv', sep = '\\t', header = True, index = False)\n",
    "df_input[df_input['positional'] == 7030].drop('positional', axis=1).to_csv('out/NSUN6_extend_structure_7030.csv', sep = '\\t', header = True, index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Getting Base Pair Probability\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "scp -r jdevans@hpc3.ust.hk:/scratch/PI/tgx/jdevans/RNAFold/Test/'ENSG00000188157(AGRN)3070_dp.ps' C:\\Users\\joele\\source\\repos\\MLRNAProject\\UBS-SeqDemoTask\\out\\Test.ps\n",
    "\n",
    "sed \"/^start of base pair probability data$/,/showpage/p\"\n",
    "\n",
    "\n",
    "sed -n \"/^%start of base pair probability data/,/^showpage/p\" <file> | grep -v -e \"%start of base pair probability data\" -e \"showpage\"\n",
    "\n",
    "sed -n \"/^%start of base pair probability data/,/^showpage/p\" ENSG00000188157\\(AGRN\\)3070_dp.ps | grep -v -e \"%start of base pair probability data\" -e \"showpage\" > temp.txt\n",
    "\n",
    "sed -n \"/^%start of base pair probability data/,/^showpage/p\" ENSG00000188157\\(AGRN\\)3070_dp.ps | grep -v -e \"%start of base pair probability data\" -e \"showpage\" > \"${file/dp.ps/prob.txt}\"\n",
    "\n",
    "\n",
    "scp -r C:\\Users\\joele\\source\\repos\\MLRNAProject\\UBS-SeqDemoTask\\out\\NSUN6_neg_extend.fasta jdevans@hpc3.ust.hk:/scratch/PI/tgx/jdevans/RNAFold/Test\n",
    "scp -r C:\\Users\\joele\\source\\repos\\MLRNAProject\\UBS-SeqDemoTask\\out\\NSUN6_pos_extend.fasta jdevans@hpc3.ust.hk:/scratch/PI/tgx/jdevans/RNAFold/Test\n",
    "## Current Verison\n",
    "\n",
    "#positive_data=\"sample_structure.fasta\"\n",
    "positive_data=\"NSUN6_pos_extend.fasta\"\n",
    "RNAfold -p < $positive_data\n",
    "for i in *_dp.ps; do\n",
    "    sed -n \"/^%start of base pair probability data/,/^showpage/p\" $i | grep -v -e \"%start of base pair probability data\" -e \"showpage\" > \"${i/dp.ps/prob_pos.txt}\"\n",
    "done\n",
    "rm *.ps\n",
    "\n",
    "# New verison\n",
    "RNAfold -p < sample_structure.fasta\n",
    "sed -n \"/^%start of base pair probability data/,/^showpage/p\" ENSG00000188157\\(AGRN\\)3070_dp.ps | grep -v -e \"%start of base pair probability data\" -e \"showpage\" > temp.txt\n",
    "\n",
    "line=\"1 22 0.003403837 ubox\"\n",
    "line=\"${line/ /-}\"\n",
    "line=\"${line// /,}\"\n",
    "\n",
    "\n",
    "sed -n \"/^%start of base pair probability data/,/^showpage/p\" ENSG00000188157\\(AGRN\\)3070_dp.ps | grep -v -e \"%start of base pair probability data\" -e \"showpage\" | while read line \n",
    "do\n",
    "   # i j v box\n",
    "   line=\"${line/ /-}\"\n",
    "   line=\"${line// /,}\"\n",
    "   without_box=${line%','*}\n",
    "   echo $without_box # Should look like i-j,v,box\n",
    "done > temp.txt\n"
   ]
  },
  {
   "attachments": {
    "image.png": {
     "image/png": "iVBORw0KGgoAAAANSUhEUgAAAUsAAADuCAYAAACwESV3AAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAADsMAAA7DAcdvqGQAACqiSURBVHhe7Z1NaB1XmvefxImdSMqim1ZHpjHySCi8IXolMt0LjyBkoywSQ6A3IouhIYsQmCYLT+ieTROGnqwCxjDBMxgPeMhipse7acbuhcWACXi0SMdIo0AaIxN10kSJhtDQ8s0kbz7e85w6dW/dW1/nOR9XVaX/D4R1y+cePVX31HNPnVN1fvd973vf+5YaxNraGm1sbJhXw3z66afmtzxLM8v05tw0zU2YDb0ere/fpQt39mjLbBpmitYW5umns+o9Zsvd/V26uHOHrh6YDc7k66bePl26u0Ov7+UrX5pZoHMq9tWJNHgViyp/cWszQCyGqRm6srSo/oZ5zfGo+l8/rPoty68tr9L5afOijP1tOrW5Z16o46nqfn5+jp6Znhh8tqr+GyXHn5G3nwy8L/OPql969JvC9iNrD0Okdfc+kcViWz4wT537nF568lu6fXmCLrxtNnaE+82/rYZPqGuLmYbOqMSzOrtI15ZnzIYsU/SLlTN0Ptt4FXPTs3R+vqi8hOK6aWKaXl48Q1dGq1cn6bXF2aFEycyp8ufPrObLuzC1QDfPZBITw/EcVv2B47nbyyQcPp6q7pcziZLh48nH/+bClNkyQN5+sqjPm5P+tPqyUz+Pma0DhO1hCJX4+Djpuh8x26qQlgcS2p8s1Yn3U9PzWN/eoFPr6/rn7PYu3eWN03P0i5HzY235DL2sTwzVe8i+Z2Ob1nu6iDsz8/26LxXEs7q4TGv83xm4R/uq6k2nZU+tb9Cr+0kgReVl8Mk8m5yoqgd2Nq1/19Q/t0BL+jdXpPXLyl/dTI/J6I/6rHSJHt34eLh3xr1IPp5J3cnP2e19/X9zs0vD7cGh/WRZWlhSn3dP/U2zYRSH9pDASVYlPvOqHml5IKUDydL0INSJ92LmkmZr7w5d1OfHBM2Xnhy3ht9zsEcvqksXH9YeTSq/u7s1dInF8TytT9hpejbbm9jbpKc3+dIte8IfqCSxRZf0CThJj1WcrLVMnaRn9MmqEoi6VE0uy1T9d0z9E7N0zqE310daf6h4Zh5NEoO6VB66dOfjeYuHLw6GLkG31PZXC9uDsP1kUW3pzdkJ1Y626IbZNIq4PWg48SVf6Hd3d82XQhXS8sCF9ifLA/Wtbn61Ij05ert0wS8vVnL3T9kzeJi5RyTZr+JktWDppLn82/+EruotKQf0a9N7lcUzjLT+MPGo5DCXJKH1Tzw/RGn7ybA2r3rIlu1I1B7MF8pdlcCfvvMns7ECaXkBPAb51ls9OveU2ZDy1Bd6+1tvfEWnzaZRTqsyb7yVvP+tt/6Xzr1QPj2iy76RluV6Vfmnisp/Sy+k5c59bbYZ0pjU33qhLCgPOpAsP6YbfI5NL9KVmcxJOWN6kCONeemRSf3v3f2Pow5+V53wcxN1yYCZonndA/Pj/5ix0KFxPcPWn+7pf+3iKUZaf5B4Mr3T31jlSp5gWTETRSPvEbafPjPLqj51ab11x6odidrDgep1bmyoKw6rnZOXHwc/+oJ++dLXNKOSW8I39ORzKsmNJjjF6Rf+Nyk7k0mOM6r8S5/TG7ny99Gv/uFB2lP/0pNfZJL413TupW/0b3vXT9CvPtC/BqUDEzwH9PqtZIxvdfEMfbi6qn940oT2d+nsreHG3D9Z1Tc9nxA3V5LyH66u0M3lGc/xO6Krn+hrt2RsbOTku7mY9IZsWFqYM+NPPdop75RYU9WzCYG0fp94ynunGXgSybSFD1d5gmXCjA2rS3RTJEHWfhJm6Ap/lqNDAAU4t4ehYRkLpOUj8+STX+uk9dpPJugn6ue1y8dMgvt/w72+01/RXz2XJMm96w/rstnyM6PlmQ8eoJ9fTlLXky9xL5J7m1/Sk5yYb5+gn/9K/Z0IdCBZKqam6NnJJAlmmVPbni/7Qn80mYUezICqy3PVu7i24jnhsbdjxhon9Gxn9uS7u5+cOLWYsTBGj3WFOg/4thKOZ8V30qgEaf1O8czQOX1sVK9uR9aT4vZQiLD9rC3zREoy5lpLiPbQRnTSOkZpB++Dt0/QP1znJPYN/fmZQQ/y9Jmvkt7nSJIrK99H/f/l28n/P/fLz+k53Ss9RpcvHNP/HYMOJEtzu8REwWyjvh1lpXA2c3V6cmj28xTPhPN/TMzSmwW3l9gz6KkMxsKSWfcXPzEvq+Ae0ZnBbPHTd5rVYzh00omd3j79uurQ8KVp+tmub9DZ3X3dHs7n2oOw/ejLb/4S2ynv1Q7h2R5ayt7H+dTywR+SZDhzMrlcziItz7x94SG6vpcm2Pvp+msnKOatna1Plunlavlso/pGL7h3kstnZz+JZ8J1edWj8BjDS+DZ7FuZkzWZdU/HS0vhnlYmUZ4NOAalx8x4Hzkeni0220Mhrd81nrV0dlk05nxAW3c26RV9e9IEPXNy8PnK2o+ZWOrt0iuiLzHH9tBiPjaJro5TJ5Neo235Ye6jjXfN+/YeoI0I45RZWp8ss2OQNrzf09dE0cfw8kzR89Mm1oLJjSRRJvfJ8awmJ8oQE1Dp/hZ9AfQnu4risURav188M/SszpX7dNGhx1054WTVHsykm7r6uGYup5Of9L5dlVjP8GubYYWa9tByTv6g4NK5gA8/TpKdbfkhnvqCfvmc6XXOfEl/VTHbHoIOXIYnVM02Ztn6WF2OqX9tywcjc3Py6E3UnAT6iXI3mdUMkSiZdH9p+tGREzhzsnp8cUjr94mnP+lVNbFTQVVPzq49HNDO/j6tF/ykl9g8DrmuLrnfN69LqWwPzaEsmZ2uSW5Fl87pe4ouuaXlh2a/b5vJoOc+z9/iFJDWJ0ub2cahb25zqwiXH71VJC3ve+/e2sJy+cxnbgZVJcpVTpTJOFbwMcr01hiapvP92X6+lYafPFG/Vt0n2J9RLh731Ujrd45nkEyrPp+15RW6sqDqHYp3Sn0Gy/1Js+z7Ze3ngF7f3KQXcz87Zp9U4tvh18Mz6LL20Cz644bPfUlPmVnp008NZrBLefILeuOFr/v3YPJ9lMl77qd3NwaX3B9sPNC/DeiNTM+Qyye9xuHyCSOz3xfSySCeHf+CYuXLDiykMXh6oRB18uVu/1AnzodpYx2lqLyQ0sUf+DG8kfG5pYUVumZO4jLWt9fpxfL8UE920miEqrqz+6F7vGWJXFq/Szz996hjuF4+xlm78Mb+8KIbTu0nR1pHjy5t3MolP0l7SFD1Lc/TvP59kubUl8Qcf5nuJ8MI+UUypOUlqB7cWyYxZbitenN8exDtHafXfv5Af9a7v5BG+v8j7OnkNjxjre+zTC+nR1Hlf1Ja/hhd/slgUif926MxhaIDl+FmtpFnO/W3e4pqLLvbxQ19b5NObezSuhk/S+glY4WeiZK5urNNl9Sl2IC0bvuJjKDom5ZHnnvnVW82qpNw2utiKi/VpfU7xKOfllH/1s1CX91M2sLwZ8u9Q1U/z3bnJs0c2o8QeXuYonmzMMeqTnzMhHldtEiGtLyEY3ThtQfpdmbW+fblh+nCO+ZlGe+cGNxbqVHvu/5wLlEyH/zqoaRs/28o9pK/M5ooB+OU+dnvty8cp9v89yKNX3ZmiTYQC3UJqxKKb8IAoO10ZoIHxAKJEgAGyRIAACxAsgQAAAuQLAEAwAIkSwAAsADJEgAALECyBAAAC5AsAQDAgiPnDa99HI7JPRInZYyeaFu43iPhDU+ftS/B5rNNPwMq8oDL6x+rxzyNHd7w4HSiZ8knlLv3OY/fklnJc8Lj8URbop+rziQmRi9s20FvuDfq86v0gAtRSW98HvNI7Qdo2p8s1Ykn8T67eKhFjM0TbQuf/GbRCn4eWcfDz0Inzyp30RuewAtuFLyvpldZ6wHvY1//eDzmsdoPSOlAsjTf2OrEE3ufs5R5qIWsjcUTLSCUp7sMaf2h4gn0eQ2hElWdB1zMWDzmEdsP6NP+ZOnhfR6gGlsoD7WhapWe3EKzEb3P8IbbI/GAR8OlPUdsP/CGD+hAsnT0PmfJ9HbsPNT1VJ3wOaVCRO9zEE93BdL6g8Rj9XlN0/mM+uHmyvACvDlmZB5wcf1DBPaYR2w/zsAb3kQG9jx77/MwVh5qS8bmiRbio46wQVq/Tzwun1fVhIo6M6094GVU16/gSa1+cg3tMVfAGw5vuBUu3vA+7h7qQprsiebbSjiejnvDd3QSGkymnFK9Lk4+TG5CRSHygCuk9ZcRymPeSHTSgje8YZjbJYTe8D62HmprBj2DwdhTT89sdtkTPTZqP689en1z5N5I1etiFW06oZJV4co94ML6U/hSOU2uIT3mDQXe8Abi6g1PWXPyUNeRnDxN8kTrMdQj6w1P6A+R9MdEzURRT+oBLyZffxmhPObNBd7wBiL1hg/j56GWcTieaHjDqwjpAZfj7zFvLvCGN5j6203y+HqoRRySJxre8AH9Xmk/GQf0gCvy9Vfj7zGPD7zhA1qfLGXe5yx2HmoXGuWJhjdcoepX25NbdbJfVgdiD7is/vLyYTzm8YE3fMDR9IYznAgsPNQuxPdEC+nva56j5g2v3I8+aZvq5Tzg0vrryvOTOsOPSLq055jtB97wlA5chrt5n2091C7E90QLcfB0M2kvh6m8VJfW7xCPtTd8p6gtqPfxF5Wqvz5RViOtfzwec3jDGXjDM8AbfhjAGw4A05kJHhALJEoAGCRLAACwAMkSAAAsQLIEAAALkCwBAMACJEsAALAAyRIAACxAsgQAAAs6clM6P1vcJE+3fTy1j8MxuUfiHOD9bKE3vJD0M6Mir3dCTI88P6d9TtW9alYIYviJnIsq/qJYcljEP0RNeRcveSzgDW80ybO0zfF0+8RTjPfCCfq56kxiYvRCsm30hqvjW+P15gQYzSPPHvDF2aFEybBW4nyg+IepKc/xqGMp8ZIDN9qfLJvm6RbGE91jrk82s2gFP5+u6+Znj5MHj5vuDR+l1uutEnFsj3zizhnUreM3awGUt7cEey95gk355Ll0Sy85cKb1yXKtYZ5ut3gKCOXFDuXpLkNav088KhHWer1je+TZA661EtmNvDK+iZ8m6bGy+m3iz2JTXuolB850ZoKnalWc3EKq5oSN4VlOEcWTQyXzQF7sLnnDrbzeh+6RL09OUi95Ezzm8IYP6EyyrDrhcwqDMXiWRfGMYuXFtqMz3nAtFuvVe70PzSNvFBVl2MafIi0/BE8wlnjJxwW84c3DZmXpQiJ5lp3jyRDSY55SuR5lAKT1y8rPCLzejt7tDC7Hv6+8oB7t5GKUxM9Iyyt40izjECr3ko8HeMObSNM83d7xBPaYZ+FbUDielnnDpV7vsXvkzdgio8eqRxKcNH7x/pZQ6iUfBzppwRveMJrm6faMJ7jHvOXoy1FOQrYr2qsErG9LUsludDZc354U2CPPPbpUkaHHwEfeJI1fvL8Gay/5eIA3vLE0zdPtHs+ahxe7Cj2G2jpvuJlo6dl7vcfqkeeecSZRns31BKXxy/e3mHIv+biAN7xVZNSqvjd1B8EmnvAe83Z7w+Ve77F55HWiTO7T1X4llSjzyVUav3x/qwgxgecKvOFt4pA83aVYxBPDY95ub7i717vqboQy7I9/JlGyzbEwUTLS+MN6zENcWcEbPqATybJRnm6FWzyDZBHUY95qb/iB2Osd3yPPk1KcKJNx6GpbpDR++f5KveRS4A0f0ImFNEoXQ+DHwArHw6aierrl8Sj6EwWqTGCP+dAkxAit8YYPoT6/Eq/34P/My1FUMi68fcjy+C8trNA1k4TKqN+HqviLKC9fuxDIvu8iLPCGp3SiZ9k0T7c8HtXoI3rMu+ANt8fcjSDybkc+/hGRe8mlwBueAm84qAHecACY7k7wgEAgUQLAIFkCAIAFSJYAAGABkiUAAFiAZAkAABYgWQIAgAVIlgAAYAGSJQAAWHAkveH8nLCX97kWWTwJ+ffwStcXbbzSNvAqOa6ebhuk9fvEw+8N6d0WtgdJ+drHEZm6RxIt9tfWkx4beMMbTfLcrLWnWzUqP+9zHcJ4NMXvmZuepfMVay9ao597ziQmRi+E201v+DA15aXtIUL7qV4ir35/OSGH9KSDYtqfLB284T7e51oc4llbThd+SFay6b9n9PlpJ/hkM4tW8PPp6f521Rs+gpV3W9geJOV9vfChPenAndYnyzWpp9vH+2yBOJ6hxp6sqJ6yxSuJq8soL3w83TZI6/eJRx2rKN5tSXsI1X5svPA28YfypINaOjPBU7Uqjmwh2DCNyzqetLFXrSvpwZHzhmcI492Wtgeb8qp3zfoIRdVak1bxB/GklwNv+IDOJMuqE95uWf0a77MQ23j6KoXAzp2UI+cNT5GWzyFtD4LyNl5y2/hDeNJjAG948wjh6Waqvc/2SOPJOmN0mZXUu7JCN/sriftTuR5lAKT1y8rPxPdujyBtD5Ly9V5ySfwDm6irJz0G8IY3kRDecDM2xBR5n0W4xvNoMss6mNFUl+eqt3BtxXcCJgPfgsLxdNwb7u3dlrYHUfl6L/l4PemR0EkL3vCGMfhmHYzdJLPKVp5u1dD7ioP9Au+zGLd4VqcndZn+bCnPhPN/TMzSmwuH1eIbgL4c5SQU2budIm0P0vJ1XnJx/OoLR9+GVXD3hb4dC97wUHQgWTI8G+ng6eaeTaah573Prsjj4R5JdjZTO7W3zSW9x5hiih5D7bw3XFp+BGl7cGg/a5Vecnn86eV/+d0X6gonxL26QuANbxUZtWrRTb+6ofPlDjfcpKHHHdspjif1aMcaUzxa3nBp+QzS9uDUfuq85PL4/Tzp8YA3vE1UerozDb3S+xyQknhSj3bV7LkPR8sb7urdlrYHt/ZT7yUfryfdBnjDB3QiWco83aqhW3uf3RDFY2794Nnz0Vs/0vd4ecSPlDdcWp6RtgfX9jP4Mij/POXxu3vS7YA3fMCR84aH8T5XI4lHM7NMH6bJdBSVPLxv/8hOQozQPW94EeXlpe3Buf3091m1AbEXvmp/0/8zL0fxbj/whqd0omfp4umOiTievU06tbE74n5O3+OZKJkj5Q1vJvG85ObuC6En3R54w1PgDQc1wBsOANPdCR4QCCRKABgkSwAAsADJEgAALECyBAAAC5AsAQDAAiRLAACwAMkSAAAsQLIEAAAL4A03HK43XFreAV4lx9XTbYO0flF5l+OTf0+lh90yntgecHn9qi79rHoJdbEEBt7wRpM8G9teb7i0vAP6ueRMImD0wrAh9lchrV9U3uX4FL+n1MMe+PhUL1yhYqvxgNfhszAGcKf9ybLt3nCH+GXwyWkWreBnzdP9bYs33OH4yDzssnhie8Dd6+cFOgreN8ZeZddpfbJca7k3XBy/lJZ7w8XHR/USRR72UMcnlAe8DJv6QVQ6M8FTtSqObGHUMXvDDeHiH6Yr3nDr4yP0sIc5Pqp3GsoDXohd/TGAN3xAZ5JlVYMuUhjkkXqiq5HG4x9/MV3xhtsen76awtLDHuT4hPSAF2FTPy+mnFFR3FwZXoB67MAb3jza7g0PFX8dletRBkBav2156fHJOml0GUsPu8/xCesBz1Nffx6esOQJsJuHZAaFN7yJtN0bHiJ+W/iWFa6/Td7wcXrYnY5PBA/4EPX17+gJy8ykzsZgwlJ/yRxGvtRJC97whuHm6e6jbxsZzIb6O3mk8XjG33ncjs/YPOzBPeAj1NWv+l+v6wlL85I54AnLW+qY8YsJeubk+LMlvOGNJWkc7fWGO8YvQI/5tc4bniI/PnyFkJ0J139rO+mJlo6JOhyftcAe8FGq66+mP4RRN+YaAXjDW0Wxp7uPTpR8ecQNMUmU4oF3ETXx5JCWL6bd3vAqio9PWr/tGKRfPOE94MPU1d9c4A1vEyWe7oRMojxkb3gp0vIltNsbXkHJ8Unrr5o9z+ITT0wPOFNffzX9XqnHlyG84QM6kSxb7Q1XSMuLaLU3PCGqh935+AySaUgP+ACb+tWxWV6hKwsq7qHPR8WvtifPmPt92cIbPgDe8ALs3NXlSL3h0vJispNYI1Ttazauw/SGi4+P1MPucnz671ExBPWAGyzrLz02hsrPzQp4w1M60bNsuzc8evwt94ZH97A7HJ94HvAE2/qv7mxQ3hmu3sdfJCp+/ysneMNT4A0HNcAbDgDT3QkeEAgkSgAYJEsAALAAyRIAACxAsgQAAAuQLAEAwAIkSwAAsADJEgAALECyBAAAC46mN3xqhp6fn6Nnpo2vRcFPPNwI5umWxcMszSzTm3OqfPokZq9H6/t36cKdQIt88CpLTfKGp/D72KHd+6R6X33iT/8G5T3dhVSUr3u8ULM/7OqWtTf1t3094NL9DQi84Y0mec5W5A0/s0gvZxouE24ZfmE8Cj4Bry1mEiUzMUGrs4t0bdlGLViDfs44k2gYDy92Duf61UnN79MO7UfMtgK84lefh8jTLS2fZ2iVn+jtbRT/+EEx7U+WDl7p5LnZDeOITsubhVJ9l+GXxqMSwUDdmi9P03OeWgA+ecwiEfxstK6fnydOHiYevzc8Rb1vRZ3U5lU5fvHXebpHieH1dmtvvIBGtn7zU9OrlO4vsKf1yXJN6pVmbzgvwHAw/BjfltqeLsPvo8IVx5OqW1UiyK7szeUvBognmBe7DKf6OVEmC+De3d01SaYEn/jVF5HI0y0tn6XM6x25vQ3hEz+opTMTPFWr4tguBBsS63gOVC/A/BqDMF7scpzqNwmQVwF6+s6fzMZifOKXerrb6PXO4h5/OfCGD+hMsqw64YuUAcPwhEy6WGqVm9ke63jSxWenF2l0sVp9ee7Z+JvmDdfoJdGSFerrcI5fi8J69p5uafksVl7vLDbtTegB94k/BvCGNw9n7zZPGvQbI0/ITKieDitF/daQlMczsBeujqheScVTuP6iA1aqBg/E9avLUgmy+meEnm5p+WGsvN6e7a16Qsgv/hjAG95EAnq35yaTXowXLvFMTdGzBX+b43m+ojMhhm8p4Xia4g2XYlm/1NMd2+tdRll7k3rA/eKPhE5a8IY3DEfvNl8K9mcZN+gsrzatvr3Pn6nwzVghjUclAH1bTMHsub49xjeeI4a+HFU9UdsVzKXlR6n1ehus25vQA+4bfyTgDW8scq/0MAe0dWeTXtG3o4SQ0tvHkxr8ymfPVQ913me6OkGPoTbIGy7Frn4z0dKz9XRLy+dZc/J6u7W3vAfcP/5YwBveKjIqU0sVaIgJj3KK4+lPYEQaU2yaN1yKrH6pp/twvd7+7c03/njAG94mHLzb9j1RB2riqZo996Fp3nApsvqlnu7D9XpL21u/F9v/cvCL3wZ4wwd0IllKvNJlnmX9bLYeqPe/V04Sj83suU/PzN2LrejP4Ab0hksR1S/1dB+e17usvck84D7x2wFv+IAj5w2vXQhh32Khghok8fCJkD7NUohKBt63D3HSk3qxFdn9COoN531enqd5/fskzamkM6dO7PX95LI0t6iGY/wD0mPcK/d0D2FRvh+T+kw9vN6j7a2uvJ0HXLq/VcAbntKJnqXEK311k58rVpcmQ05p7r3tJ7PRAW6/kHmuzex5zv2skscuvyfAfZYOXmwmmjdcnczzZqGHVZ0omYn+4g+5RTUc44+Jtddb2N7ie8ClwBueAm84qAHecACY7k7wgEAgUQLAIFkCAIAFSJYAAGABkiUAAFiAZAkAABYgWQIAgAVIlgAAYAGSJQAAWNC9m9JTZ3Kdh1oEP4ss8YBLyzvA++nq0a5FEj8vyMuLz5ZQ9vioNP62l7ek6NE/XkG8bPXvLnu6m0bHepaqAdt4qEUkz9nae8Cl5R3QzyVnTlRGLxQcwgPewPjbXh50gg4lSz7JVQM2r4Ih9YA7eMxlqP2M6QF3jp8XlEjKDv3kepXS+NteXgY/J9330FzHKFmT6MinoRqwXmWFFzao8VALWRN6wKXlxUT2gDcu/raXB52hG8nSNGAbD7UrVavuFC3cKy1vS2wPeEpT4m97+XHSNU930+hGstRLeNl5qF2pOgGKtADS8rbE9oCnyOOftvJcS+Nve/mx0UFPd9PozqCI0ENti9QDLveGu1G5vqQHoeKv9lzL4297+dh00dPdNLqTLGMh9YAH9JjXwrevcP0hPd0O8Us9132k8be9fEx00uqWp7tpIFnWIvWAS8s3DWn8Qs81iEIXPd1NA8nSiuTkt/eSS8vL0WOK0TzdYeLvX9IXjOFJ4297+dh00dPdNJAsnSn2gJcjLV9MbE93OYcTf9vLj4suerqbBpKlK1IvubR8CbE93aU4xL+W81zL4297+XHRRU9300CytEDiAWek5UXE9oArQnjYiz3XCmn8bS/vQZGzu5QOerqbRkcW0piSeaiFyDzg8vJi9LPJ5pG7EUJ4wCXxO3mupfG3vbwzJc7uzKIZXfZ0N41jExMTf2t+bwRPPPEEffTRR+bVMPfumeSX47v04v/9M1qdVIly8kH6jt72oPqdX6uf45/Tv3z4GblORr93r0eTJ6boR6ruBPaA/47+8p0d+k+zJYu0vJgvP6N//rRHS9/9Ps31/8Q+Xfrtf9Grn5nXBbx33yP0199PxtV++9H79O8lPVxJ/O/d+5T+8PUJWjg+Sd9JiyvYc/26iudne1+aLRmk8be9vDP308bt++nE49/SwtQgYe69+yBt/D75ffYvvqIfnlTb/uNh+sd3v6HHf6hyuf4f9nQ/RH/3T/mLxz9uP0C3/0eV/YEqm14RsKf7Xx+iv/m3kfKqx/n3LyQ9zuuvPUTX/5hsZn6/cT+d/vE3dHLqa3r8xAN0Y3u0R9ot4A0/csADDoALGLM8ciBRAuACkiUAAFiAZAkAABYgWQIAgAVIlgAAYAGSJQAAWIBkCQAAFiBZAgCABZ24KZ2fXT43N02rZsl/hp8gubi1ObzOojP87K/MA740s0xvqpjm0pB6/Pjl3XAuc154NoK3OkGyv7wAboVVcx/ecAnwhjeX9vcsVVLiVbyziZJhrcH5IB7nxBwp8Wjz89LXFjOJklHxrc4u0rXlAOo//Wxy5kRlgnmr5fsrRhp/28uDTtCJy/C7WmswcFxrj7PRGnh7uqUebXUi/dQsLMGri4+Wp+m5yhV/6onrrYY3PHR5GfCGN5f2fxp7m/S01hpkr394pW/jcaZJeswjOa1JPdpTvOKRQp1IvLp4Cpe/aDQL8z7JMrK3Wry/Uprm9Y5dHnSGI/DV5ZmcDFULumrFQMpB1l0TnnF5q633V0jbPeDjOv4uwBselw4nyyma1z2AMFSdAEOKgXRx2OlFujKygK6+PPdcHHZc3mrr/e0zDW94wOMvBt7w6HQ2WS4tzJkZ2h7t5Nu1NXKP9sCOuDqikqX9XTp76465dPMjlroA3nA7Yh1/V+ANj083k+XUAr05a3oAPPbm065dPOBTU/TsZPL3s8ypbc+H7HTw7SscT0hvNbzh8crHRCcteMNj0r1kqW/rGMxWlqkT7JF6tNUJpG8rKZhN1reXVPtvDh94w9sIvOHx6Vay5G/6TKI8uxlKhpKc/DYe7fTyv3w2WfXY5v2nS/WYYjRvtf3+VgFv+PiANzw+3UmWOlGqHp369a5JlCHGBsvJqE8zg/39CYBIY1qH560u3l8pbfeAH97xrwbe8Ph0JFlmEiXbBKMnSkWNR1v3PCJwaN5qeMM1h3b8a4A3PD4dSJY8yM6JMhlX8x+jzCPxaNvMJnv1PHy81TyeyxMS8IYPiF3eA3jDm0XrF9JYWliha2bmuwxfl7PMAz6ln63WJ04R6mTyvn0oO4k1QtW+wht+SOWdgTe8SXRnzDIiV3e26ZK5FSaBPdrbKukVDeyb2eRddbmWfQv3fHf5PQHuszy4Q09vbNP6UEj7dGmj+kRNe71M1aWiZH+v7hTtK/eeVWJV8RQmZGn8bS/vzDG68Npxur5ncZq+c2Jwb6WGveEP5xIlw8+f67L9mW0Fe8MvP5xLlINxyvzs99sXjtNt/ntHZPwS3vAjB7zhALiAnuWRA4kSABeQLAEAwAIkSwAAsADJEgAALECyBAAAC5AsAQDAAiRLAACwAMkSAAAs6IY3fGqGnp+fo2emjSxMwU+Q3KjwesvgZ3/tPNp1j/9p9ktc2hJ4laUI3uoE+/0dPJtfQtm+SuNve3lL4A1vLu3vWbI3/MwivZxJlEyd1sCe5FnvkB5t7yW89LPJmROVCeatDr+/OaTxt7086ASduAxPnkPeMA7n5Ofsdmb1H598KfRoX90cxDD8s03ruoT9EmfFxPVWwxseurwMeMObS/s/DfaG8wIPB8OP8W2p7anWwEeFuxbKoz3zaHKpOrLEmZjI3upg+1tG2z3gkY8/aC746rKkapWe+oV+VW9kLklC65/4jVWOy1vtt7/ltN0DPq7j7wK84XHpaLLkCYp08dl9+k2AZbOqToAixcAQmd6Ibyzj8lbL93ca3vCAx18MvOHR6U6y5EH3/snKExQTdFcrWovWnLQnhEe7vDfiTix1AbzhdsQ6/q7AGx6fTl+Gs6fbGxdv+BAzdE6v5N6jSzsBurhZ+PYVjiektxre8HjlY6KTFrzhMelOsuTVq9OTdX2DzvLq3ap3c97b0y31aI+QTuz09unXzeqMlABveBuBNzw+He1ZHtDWnU16Rd/OEeJkTU5+F4/2Wjq7vP+xmTkNgx5TjOatdt/fLPCGjw94w+PT6cvwuAPuGfVpwWB/wgw9ayaZLgayTh6et9pmf+tpuwf88I5/NfCGx6fTyTJtvFGw8GgvLcyZeyvDTewcmrca3nDNoR3/GuANj0/rk2WZt3ppZpneNIpc33sbJR7tAYOTx/fvD+Hjre7fMQBveJ/Y5T2AN7xZtH4hjdqFK/b9F62QeLT76OeH+bE4VWY98JhWv+48Vd7q7H7AG54hdnln4A1vEscmJib+1vzeCJ544gn66KOPzKth7t1LxiCzvPfJp/SHYyfo2PH7aO7BB81WdZKqE/vq7/6bfrzzmdniznv3ejR5Yop+NJnWzx7t39FfvrND/2m2jLL2+J/TC5OcLP6bfvbZl2ZrIL78jP750x4tfff7ap/NNrW/l377X/Rqxe6+d98j9NffT4YmfvvR+/TvxblStL/v3VPH/+sTtHB8kr4zOPz6+L+u4vnZXsG+S+Nve3ln7qeN2/fTice/pYWpQcLce/dB2vh98vvsX3xFPzyptv3Hw/SP735Dj/9Q5XL9P+wNf4j+7p/yF49/3H6Abv+PKvsDVTa9ImBv+L8+RH/zbyPlVY/z719IepzXX3uIrv8x2cz8fuN+Ov3jb+jk1Nf0+IkH6Mb2aI+0W8AbfuSANxwAF1o/ZgmkIFEC4AKSJQAAWIBkCQAAFiBZAgCABUiWAABgAZIlAABYgGQJAAAWIFkCAIAFHbopnZ/PHXZd80rpF3dG1lp0Il+3fmLD1kvOi8TOP6p+6dFvgsSj4DojeKsTJPvLC+CqOMyrHGWPm0rjb3t5S+ANby4d6VlOFbqu56Zn6fy8r2qvuG7tibbyaLM6VZ1U09P65zGz1Qv9bHLmRGWCeat999cCafxtLw86QSeS5drymWTFF9Vz4xW9U73B2Y1tWtcrxHjg7NFOWNKr0fTorm8cfTj5mkUcInir4Q0PXV4GvOHNpf2fhvqW/6lZ9WZ9O1nRO2WLV7K+47cEzJqPR1vFxsvErW9v0Q2zyZvI3mqv/bWhaV7v2OVBZ+hAspxIvuUDriNYRNWCrloxUMDavOqBBI5rXN5ql/21oWle79jlxwm84XFpfbLsL+Uf2HEzStUJUKQYoJllOj+tLmW37gSNa1zeavH+8mK4emHh5Afe8DEDb3h0Wp8s+41X9YT0it4r6Qm7Qjf7K1m74+bRnqEr/H+lq6j7E0tdAG+4HbGOvyvwhsen/ZfhKY8ua7f1XJI7FeryfHqRrq34Dbi7eLTXlhdp1YxpRYVvX+F4Qnqr4Q2PVz4mOmnBGx6TziTL1enJoZnwUzwTzv8xMUtvlvRu7BB6tPXlt+p57O6MjGm1BXjD2wi84fHpTLLk2dvsTLh2Om+bS0rvMaTk5K/3aE/RL+ZUpuzt0iuB1Ldl6DHFaN5q2/2tBt7w8QFveHxanyxTj/P4x5Ay6tP+YP8UzfMm1Zu9xpdn/Z/0PlB1aXuGX7tfth2et7pof+U0zesdu/y4gDc8Pq1PlqnHWX/Tj5NCj/YB7ezv03rBT3pJy+N+6+oS933zWsqheavhDdcc2vGvAd7w+LT/Mtx4nHki4UrJ7O34vOEH9PrmJr2Y+9kxrmmVaHb4tcftRD7ean5MT/d04Q3vE7u8B/CGN4tuLKQxs0wflt3Wohrv2Vt+9zo6ecOHmNLPW/Njj5c2bvnfTqSfTTaP3I1Q5a2GN/yQyjsDb3iTaH/PktnbpFMbu7RuxpMS2HW97Z0omas723TJ3AqTkNZ9SAP7B3fo6dHn3nnVm43qEzWdcGGqLhUl+3t1h5+LVpem2eIK9obz7USFCVkaf9vLO3OMLrx2nK7vWZym75wY3FupYW/4w7lEyfDz57psf2Zbwd7wyw/nEuVgnDI/+/32heN0m//eERm/hDf8yAFvOAAudKNnCQQgUQLgApIlAABYgGQJAAAWIFkCAIAFSJYAAGABkiUAANRC9P8BrEvilkfvWrIAAAAASUVORK5CYII="
    }
   },
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Collect a bulk file for the base pairing probabilities using the following code\n",
    "- Need to filter out lbox data (only 0.95 probabilities)\n",
    "\n",
    "![image.png](attachment:image.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "## To Upload\n",
    "scp -r C:\\Users\\joele\\source\\repos\\MLRNAProject\\UBS-SeqDemoTask\\out\\NSUN6_pos_extend.fasta jdevans@hpc3.ust.hk:/scratch/PI/tgx/jdevans/RNAFold/Test\n",
    "\n",
    "\n",
    "## Main\n",
    "positive_data=\"NSUN6_pos_extend.fasta\"\n",
    "RNAfold -p < $positive_data\n",
    "for i in *_dp.ps; do\n",
    "    sed -n \"/^%start of base pair probability data/,/^showpage/p\" $i | grep -v -e \"%start of base pair probability data\" -e \"lbox\" -e \"showpage\" | while read line \n",
    "    do\n",
    "        # i j v box\n",
    "        line=\"${line/ /-}\"\n",
    "        line=\"${line// /,}\"\n",
    "        without_box=${line%','*}\n",
    "        without_box+=','\n",
    "        name_positional=\"${i/_dp.ps/}\"\n",
    "        name_positional=\"${name_positional/_/,}\"\n",
    "        without_box+=$name_positional\n",
    "        echo $without_box # Should look like i-j,v,sequencing\n",
    "    done\n",
    "done > NSUN6_pos_extend_bpp.csv\n",
    "rm *.ps\n",
    "\n",
    "\n",
    "\n",
    "## To collect the file run on local shell\n",
    "scp -r jdevans@hpc3.ust.hk:/scratch/PI/tgx/jdevans/RNAFold/Test/NSUN6_pos_extend_bpp.csv C:\\Users\\joele\\source\\repos\\MLRNAProject\\UBS-SeqDemoTask\\out\\"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now the file needs to be reformatted/merged (bpp is base-pairing-probability)\n",
    "- Split files by selections (i.e. bpp_3070, bpp_5050, bpp_7030)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 333,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(219, 4753)\n"
     ]
    }
   ],
   "source": [
    "bpp_data = pd.read_csv(r\"out\\NSUN6_pos_extend_bpp.csv\")\n",
    "current_head = bpp_data.columns\n",
    "bpp_data.columns = ['pair', 'P', 'gene', 'selection']\n",
    "bpp_data.loc[len(bpp_data.index)] = current_head\n",
    "#print(bpp_data)\n",
    "\n",
    "seq_data = pd.read_csv(r\"out/NSUN6_pos_extend.csv\")\n",
    "current_head = seq_data.columns\n",
    "seq_data.columns = ['gene','selection', 'seq']\n",
    "seq_data.loc[len(seq_data.index)] = current_head\n",
    "\n",
    "bpp_3070 = bpp_data[bpp_data['selection'] == 3070].drop('selection', axis=1)\n",
    "bpp_5050 = bpp_data[bpp_data['selection'] == 5050].drop('selection', axis=1)\n",
    "bpp_7030 = bpp_data[bpp_data['selection'] == 7030].drop('selection', axis=1)\n",
    "\n",
    "#new_order = ['Gene', 'Pair', 'P']\n",
    "#bpp_5050 = bpp_5050[new_order]\n",
    "#print(bpp_5050)\n",
    "\n",
    "genes = pd.unique(bpp_3070['gene'])\n",
    "\n",
    "RPUSD1 = bpp_3070.loc[bpp_3070['gene'] == genes[0]].drop('gene', axis=1)\n",
    "RPUSD1_pair = RPUSD1['pair'].to_list()\n",
    "RPUSD1_prob = RPUSD1['P'].to_list()\n",
    "RPUSD1_seq = seq_data[(seq_data['gene'] == genes[0]) & (seq_data['selection'] == 3070)]['seq'].to_list()\n",
    "\n",
    "\n",
    "joinedlist = [genes[0]] + RPUSD1_seq + RPUSD1_prob\n",
    "temp = np.array([joinedlist])\n",
    "df = pd.DataFrame(temp, columns=['gene', 'seq'] + RPUSD1_pair)\n",
    "\n",
    "\n",
    "## For 3070\n",
    "seq_3070 = seq_data[seq_data['selection'] == 3070].drop_duplicates(subset=['gene'])\n",
    "all_genes = seq_3070['gene'].to_list()\n",
    "all_seq = seq_3070['seq'].to_list()\n",
    "all_pairs = pd.unique(bpp_3070['pair'])\n",
    "\n",
    "df_pairs_probs = pd.DataFrame(index=range(len(all_genes)),columns=all_pairs)\n",
    "total_pairs_count = len(all_pairs)\n",
    "\n",
    "#pairs_with_probs = bpp_3070.loc[bpp_3070['gene'] == genes[0]].drop('gene', axis=1)\n",
    "#for pair in pairs_with_probs['pair']:\n",
    "#    df_pairs_probs.loc[0, pair] = pairs_with_probs[pairs_with_probs['pair']==pair]['P'].to_list()[0]\n",
    "\n",
    "\n",
    "\n",
    "# pairing_probs = bpp_3070.loc[bpp_3070['gene'] == genes[0]].drop('gene', axis=1)\n",
    "# pairings_unique = sorted(set(all_pairs) - set(pairing_probs['pair'].to_list()))\n",
    "# new_record = pd.DataFrame(np.array([np.full((len(pairings_unique)), 'NaN')]), columns=pairings_unique)\n",
    "\n",
    "# pairing_probs = pairing_probs.T\n",
    "# pairing_probs.columns = pairing_probs.iloc[0]\n",
    "# pairing_probs = pairing_probs.iloc[1:, :]\n",
    "# #print(pairing_probs)\n",
    "# pairing_probs = pairing_probs.assign(**new_record)\n",
    "# #print(pairing_probs)\n",
    "\n",
    "i = 0;\n",
    "for gene in all_genes:\n",
    "    pairs_with_probs = bpp_3070.loc[bpp_3070['gene'] == gene].drop('gene', axis=1)\n",
    "    for pair in pairs_with_probs['pair']:\n",
    "        probability = pairs_with_probs[pairs_with_probs['pair']==pair]['P'].to_list()[0]\n",
    "        df_pairs_probs.loc[i,pair] = probability;\n",
    "    i = i+1\n",
    "print(df_pairs_probs.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 334,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(219, 2)\n"
     ]
    }
   ],
   "source": [
    "# Concatenate along the first axis (rows)\n",
    "restruct = pd.DataFrame([all_genes, all_seq]).T\n",
    "restruct.columns=['gene','seq']\n",
    "print(restruct.shape)\n",
    "restruct = pd.concat([restruct,df_pairs_probs], axis=1)\n",
    "\n",
    "\n",
    "restruct.to_csv('out/restruct_3070.csv', index=False)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "MLRNA",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
